collectors/mmlsqos.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2020 Trey Dockendorf
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package collectors
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var (
+	qosFilesystems = kingpin.Flag("collector.mmlsqos.filesystems", "Filesystems to query with mmlsqos, comma separated. Defaults to all filesystems.").Default("").String()
+	qosTimeout     = kingpin.Flag("collector.mmlsqos.timeout", "Timeout for mmlsqos execution").Default("60").Int()
+	qosSeconds     = kingpin.Flag("collector.mmlsqos.seconds", "Display the I/O performance values for the previous number of seconds. The valid range of seconds is 1-999").Default("60").Int()
+	qosMap         = map[string]string{
+		"pool":      "Pool",
+		"timeEpoch": "Time",
+		"class":     "Class",
+		"iops":      "Iops",
+		"ioql":      "Ioql",
+		"qsdl":      "Qsdl",
+		"et":        "ET",
+		"MBs":       "MBs",
+	}
+	MmlsqosExec = mmlsqos
+)
+
+type QosMetric struct {
+	Pool  string
+	Time  float64
+	Class string
+	Iops  float64
+	Ioql  float64
+	Qsdl  float64
+	ET    float64
+	MBs   float64
+}
+
+type MmlsqosCollector struct {
+	Iops   *prometheus.Desc
+	Ioql   *prometheus.Desc
+	Qsdl   *prometheus.Desc
+	ET     *prometheus.Desc
+	MBs    *prometheus.Desc
+	logger log.Logger
+}
+
+func init() {
+	registerCollector("mmlsqos", false, NewMmlsqosCollector)
+}
+
+func NewMmlsqosCollector(logger log.Logger) Collector {
+	labels := []string{"fs", "pool", "class", "measurement_period_seconds"}
```

Comment:
I wonder if it's better to make `measurement_period_seconds` a metric rather than a label?  Then people could do operations with `time()` against that metric.  Moving to a metric would recommend still having `fs`, `pool` and `class` if those can be associated as then can be linked to other metrics.

Reply:
In some configuration cases we can have multiple values for `measurement_period_seconds` for the same `fs`, `pool` and `class` during one execution of command `mmlsqos`. And in this situation we'll face with errors while Prometheus will try to generate metric `measurement_period_seconds` with same labels and I haven't found any additional valuable label which can split such metrics.
That's why I suppose that it's better to store `measurement_period_seconds` in labels rather than in metrics.

I've attached some example of output just for better understanding:
<img width="111" alt="image" src="https://user-images.githubusercontent.com/54704017/226824748-76a5c83e-9db5-4390-961d-3856e631ecd8.png">

Reply:
Thanks for clarifying, I think it's fine to leave this the way you have it then.